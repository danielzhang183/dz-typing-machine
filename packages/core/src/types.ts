export interface InsertPatch {
  type: 'insert'
  from: number
  text: string
}

export interface DeletePatch {
  type: 'removal'
  from: number
  length: number
}

export type Patch = InsertPatch | DeletePatch

export interface Snapshot {
  content: string
  options?: SnapshotOptions
}

export interface SnapshotOptions {
  wait?: number
  pause?: boolean
}

export interface AnimatorStepInsert {
  type: 'insert'
  cursor: number
  content: string
  char: string
}

export interface AnimatorStepRemoval {
  type: 'removal'
  cursor: number
  content: string
}

export interface AnimatorStepInit {
  type: 'init'
  content: string
}

export interface AnimatorStepPatch {
  type: 'new-patch'
  patch: Patch
  index: number
}

export interface AnimatorStepSnap {
  type: 'new-snap'
  snap: Snapshot
  index: number
}

export type AnimatorStep = AnimatorStepInsert | AnimatorStepRemoval | AnimatorStepSnap | AnimatorStepInit | AnimatorStepPatch
