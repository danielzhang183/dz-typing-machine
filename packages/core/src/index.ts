export * from './animator'
export * from './patch'
export * from './snaps'
export * from './types'
