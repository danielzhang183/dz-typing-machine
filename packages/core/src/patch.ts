import { diff_match_patch as DMP } from 'diff-match-patch'
import type { Diff } from 'diff-match-patch'
import type { Patch } from './types'

export function patch(input: string, output: string): Diff[] {
  const differ = new DMP()
  const delta = differ.diff_main(input, output)
  differ.diff_cleanupSemantic(delta)

  return delta
}

export function calculatePatch(diff: Diff[]): Patch[] {
  const patches: Patch[] = []
  let cursor = 0
  for (const change of diff) {
    const length = change[1].length
    if (change[0] === 0) {
      cursor += length
      continue
    }
    else if (change[0] === -1) {
      patches.push({
        type: 'removal',
        cursor: cursor + length,
        length,
      })
    }
    else if (change[0] === 1) {
      patches.push({
        type: 'insert',
        cursor,
        content: change[1],
      })
      cursor += length
    }
    else {
      throw new Error('Unknown Patch Type')
    }
  }

  return patches
}
