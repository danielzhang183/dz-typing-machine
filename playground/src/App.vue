<template>
  <div font-sans leading-1em>
    <Playground />
  </div>
</template>
