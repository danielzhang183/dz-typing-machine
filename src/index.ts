export * from './animator'
export * from './patch'
export * from './types'
