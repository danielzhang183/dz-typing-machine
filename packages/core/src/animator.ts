import { calculatePatch, patch } from './patch'
import type { AnimatorStep, Patch } from './types'

export function* createAnimator(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let output = input
  let cursor = 0
  for (let index = 0; index < patches.length; index++) {
    const patch = patches[index]
    yield { type: 'new-patch', patch, index }

    if (patch.type === 'insert') {
      cursor = patch.cursor
      const head = output.slice(0, patch.cursor)
      const tail = output.slice(patch.cursor)
      let selection = ''
      for (const char of patch.content) {
        selection += char
        yield {
          type: 'insert',
          char,
          cursor: cursor + selection.length,
          content: head + selection + tail,
        }
      }
      output = head + patch.content + tail
    }
    else if (patch.type === 'removal') {
      cursor = patch.cursor - patch.length
      const head = output.slice(0, cursor)
      const tail = output.slice(patch.cursor)
      const selection = output.slice(cursor, patch.cursor)
      for (let i = selection.length - 1; i >= 0; i--) {
        yield {
          type: 'removal',
          cursor: cursor + i,
          content: head + selection.slice(0, i) + tail,
        }
      }
      output = head + tail
    }
  }

  yield { type: 'snap-finish', content: output }
}

export function simpleAnimator(input: string, output: string) {
  const delta = patch(input, output)
  const patches = calculatePatch(delta)
  return createAnimator(input, patches)
}

export function applyPatches(input: string, patches: Patch[]) {
  for (const patch of createAnimator(input, patches)) {
    if (patch.type === 'snap-finish')
      return patch.content
  }
  return input
}
