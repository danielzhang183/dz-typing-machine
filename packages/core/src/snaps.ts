import { simpleAnimator } from './animator'
import type { AnimatorStep, Snapshot } from './types'

export const SNAP_HEADING = 'typingMachine Snapshots v1'
export const SNAP_SEPERATOR_PRE = '-'.repeat(2)
export const SNAP_SEPERATOR_POST = '-'.repeat(10)
export const SNAP_SEPERATOR_OPTIONS = '-----OPTIONS----'
export const SNAP_SEPERATOR = `${SNAP_SEPERATOR_PRE}--${SNAP_SEPERATOR_POST}`
export const SNAP_SEPERATOR_MATCHER = new RegExp(`\\n?${SNAP_SEPERATOR_PRE}[#\\w-]*${SNAP_SEPERATOR_POST}\\n`, 'g')
export const SNAP_SEPERATOR_OPTIONS_MATCHER = new RegExp(`\\n?${SNAP_SEPERATOR_OPTIONS}\\n`, 'g')

export class Snapshots extends Array<Snapshot> {
  constructor(...args: Snapshot[]) {
    super(...args)
  }

  get last() {
    return this[this.length - 1]
  }

  get first() {
    return this[0]
  }

  toString() {
    return [
      SNAP_HEADING,
      ...this
        .flatMap((snap, i) => [
          SNAP_SEPERATOR_PRE + (i + 1).toString().padStart(2, '0') + SNAP_SEPERATOR_POST,
          snap.content,
          ...(
            snap.options
              ? [
                  SNAP_SEPERATOR_OPTIONS,
                  Object.keys(snap.options).length > 1
                    ? JSON.stringify(snap.options, null, 2)
                    : JSON.stringify(snap.options),
                ]
              : []
          ),
        ]),
      SNAP_SEPERATOR,
      '',
    ].join('\n')
  }

  static fromString(raw: string) {
    const parts = raw
      .split(SNAP_SEPERATOR_MATCHER)
      .slice(1, -1) // remove header and tailing

    const snapshots: Snapshot[] = []
    for (let i = 0; i < parts.length; i += 1) {
      const [content, optionsRaw] = parts[i].split(SNAP_SEPERATOR_OPTIONS_MATCHER)
      const snap: Snapshot = {
        content,
      }
      if (optionsRaw?.trim())
        snap.options = JSON.parse(optionsRaw)

      snapshots.push(snap)
    }

    return new Snapshots(...snapshots)
  }

  * animate(): Generator<AnimatorStep> {
    let lastContent: string | undefined
    const copy = [...this]
    for (let index = 0; index < copy.length; index++) {
      const snap = copy[index]
      if (lastContent == null) {
        lastContent = snap.content
        yield {
          type: 'init',
          content: lastContent,
        }
        continue
      }

      yield {
        type: 'new-snap',
        snap,
        index,
      }

      const animator = simpleAnimator(lastContent, snap.content)
      for (const result of animator)
        yield result

      lastContent = snap.content
    }
  }
}

export type SnapshotFallbackLoader = (id: string) => Snapshots | undefined | Promise<Snapshots | undefined>
export interface SnapshotManagerOptions {
  ensureFallback?: SnapshotFallbackLoader
}

export class SnapshotManager extends Map<string, Snapshots> {
  constructor(
    public options: SnapshotManagerOptions = {},
  ) {
    super()
  }

  async ensure(
    id: string,
    load = this.options.ensureFallback,
  ) {
    if (!this.has(id))
      this.set(id, await load?.(id) || new Snapshots())
    return this.get(id)!
  }
}
