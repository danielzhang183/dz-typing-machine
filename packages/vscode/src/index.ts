import { existsSync, promises as fs } from 'fs'
import { Range, Selection, commands, window, workspace } from 'vscode'
import { SnapshotManager, Snapshots, simpleAnimator } from '../../core/src'

const snapExt = '.typingMachine'

export function activate() {
  function getSnapPath(id: string) {
    return id + snapExt
  }

  const manager = new SnapshotManager({
    async ensureFallback(path) {
      const filepath = getSnapPath(path)
      if (existsSync(filepath)) {
        const content = await fs.readFile(filepath, 'utf-8')
        const snap = Snapshots.fromString(content)
        window.showInformationMessage('typingMachine: Snapshots loaded from file')

        return snap
      }
    },
  })

  workspace.createFileSystemWatcher(`**/*\\${snapExt}`)
    .onDidCreate((uri) => {
      manager.delete(uri.path.replace(snapExt, ''))
    })

  async function writeSnapshots(path: string, snap: Snapshots) {
    const filepath = getSnapPath(path)
    await fs.writeFile(filepath, snap.toString(), 'utf-8')
  }

  commands.registerCommand('typingMachine.snap', async () => {
    const doc = window.activeTextEditor?.document
    if (!doc)
      return
    const path = doc.uri.fsPath
    if (path.endsWith(snapExt))
      return

    const snaps = await manager.ensure(path)

    snaps.push({ content: doc.getText() })
    await writeSnapshots(path, snaps)

    window.showInformationMessage(`typingMachine: Snapshot added (${snaps.length})`)
  })

  commands.registerCommand('typingMachine.play', async () => {
    const editor = window.activeTextEditor
    const doc = editor?.document
    if (!editor || !doc)
      return

    const path = doc.uri.fsPath
    const snaps = await manager.ensure(path)
    if (!snaps?.length) {
      window.showErrorMessage('typingMachine: No Snapshot Found!')
      return
    }

    const lastSnap = snaps[snaps.length - 1]
    if (lastSnap.content !== doc.getText()) {
      const take = 'Take Snapshot'
      const discard = 'Discard'
      const cancel = 'Cancel'

      const result = await window.showInformationMessage(
        'The current document has been modified since last snapshot. Do you want take another snapshot?',
        { modal: true },
        take,
        discard,
        cancel,
      )
      if (!result || result === cancel)
        return
      if (result === take)
        await commands.executeCommand('retypewriter.snap')
    }

    window.showInformationMessage('typeMachine: Playing...')
    let lastContent: string | undefined
    for (const snap of snaps) {
      if (!lastContent) {
        lastContent = snap.content
        editor.edit((edit) => {
          edit.replace(
            new Range(0, 0, Infinity, Infinity),
            lastContent!,
          )
        })

        continue
      }

      const animator = simpleAnimator(lastContent, snap.content)
      let lastIndex = -1
      for (const result of animator) {
        if (lastIndex !== result.patchIndex)
          await sleep(900)

        await editor.edit((edit) => {
          if (result.char != null) {
            edit.insert(
              doc.positionAt(result.cursor - 1),
              result.char,
            )
          }
          else {
            const range = new Range(
              doc.positionAt(result.cursor),
              doc.positionAt(result.cursor + 1),
            )
            edit.delete(range)
          }
        })

        const pos = doc.positionAt(result.cursor)
        editor.selection = new Selection(pos, pos)

        await sleep(Math.random() * 60)
        lastIndex = result.patchIndex
      }

      lastContent = snap.content
    }

    window.showInformationMessage('typeMachine: Finished...')
  })
}

export function deactivate() {

}

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}
