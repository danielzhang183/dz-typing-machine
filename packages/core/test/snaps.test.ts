import { expect, it } from 'vitest'
import type { Snapshot } from '../src'
import { Snapshots } from '../src'

it('snaps', () => {
  const data: Snapshot[] = [
    { content: '' },
    { content: 'import {} from "vue"' },
    { content: 'import { createApp } from "vue"\n\nconst app = createApp()\n' },
  ]

  const snaps = new Snapshots(...data)
  const serialized = snaps.toString()

  expect(serialized).toMatchSnapshot()

  const deserialized = Snapshots.fromString(serialized)
  expect(data).toEqual([...deserialized])
})
