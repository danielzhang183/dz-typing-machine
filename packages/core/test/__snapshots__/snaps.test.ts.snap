// Vitest Snapshot v1

exports[`snaps 1`] = `
"typingMachine Snapshots v1
--01----------

--------------
--02----------
import {} from \\"vue\\"
--------------
--03----------
import { createApp } from \\"vue\\"

const app = createApp()

--------------
--------------
"
`;
